syntax = "proto3";

message EspHostedCmdGetStatus {
    string cmd = 1;
    int32 mode = 2;
}

message EspHostedRespGetStatus {
    string resp = 1;
    int32 mode = 2;
}

message EspHostedCmdSetMacAddress {
    bytes mac = 1;
    int32 mode = 2;
}

message EspHostedRespSetMacAddress {
    bytes resp = 1;
}

enum EspHostedEncryptionMode {
    Type_Open = 0;
    Type_WEP = 1;
    Type_WPA_PSK = 2;
    Type_WPA2_PSK = 3;
    Type_WPA_WPA2_PSK = 4;
    Type_WPA2_ENTERPRISE = 5;
    Type_WPA3_PSK = 6;
    Type_WPA2_WPA3_PSK = 7;
}

message EspHostedCmdConfig {
    string ssid = 1;
    string pwd = 2;
    EspHostedEncryptionMode ecn = 3;
    int32 chnl = 4;
    int32 max_conn = 5;
    bool ssid_hidden = 6;
    string bssid = 7;
    int32 rssi = 8;
    string status = 9;
    int32 bw = 10;
    bool is_wpa3_supported = 11;
    int32 listen_interval = 12;
}

message EspHostedRespConfig {
    string ssid = 1;
    string pwd = 2;
    EspHostedEncryptionMode ecn = 3;
    int32 chnl = 4;
    int32 max_conn = 5;
    bool ssid_hidden = 6;
    string bssid = 7;
    int32 rssi = 8;
    string status = 9;
    int32 bw = 10;
    bool is_wpa3_supported = 11;
    int32 listen_interval = 12;
}

message EspHostedScanResult {
    bytes ssid = 1;
    uint32 chnl = 2;
    int32 rssi = 3;
    bytes bssid = 4;
    EspHostedEncryptionMode ecn = 5;
}

message EspHostedCmdScanResult {
}

message EspHostedRespScanResult {
    uint32 count = 1;
    repeated EspHostedScanResult entries = 2;
}

message EspHostedConnectedSTAList {
	bytes mac = 1;
	int32 rssi = 2;
}

message EspHostedCmdConnectedSTA {
}

message EspHostedRespConnectedSTA {
	uint32 num = 1;
	repeated EspHostedConnectedSTAList stations = 2;
}

message EspHostedCmdSetPowerSaveMode {
    int32 power_save_mode = 1;
}

message EspHostedRespSetPowerSaveMode {
    bytes resp = 1;
}

message EspHostedCmdGetPowerSaveMode {
}

message EspHostedRespGetPowerSaveMode {
    int32 power_save_mode = 1;
    bytes resp = 2;
}

enum EspHostedConfigMsgType {
    TypeCmdGetMACAddress = 0;
    TypeRespGetMACAddress = 1;
    TypeCmdGetWiFiMode = 2;
    TypeRespGetWiFiMode = 3;
    TypeCmdSetWiFiMode = 4;
    TypeRespSetWiFiMode = 5;
    TypeCmdGetAPConfig = 6;
    TypeRespGetAPConfig = 7;
    TypeCmdSetAPConfig = 8;
    TypeRespSetAPConfig = 9;
    TypeCmdGetSoftAPConfig = 10;
    TypeRespGetSoftAPConfig = 11;
    TypeCmdSetSoftAPConfig = 12;
    TypeRespSetSoftAPConfig = 13;
    TypeCmdDisconnectAP = 14;
    TypeRespDisconnectAP = 15;
    TypeCmdGetAPScanList = 16;
    TypeRespGetAPScanList = 17;
    TypeCmdGetConnectedSTAList = 18;
    TypeRespGetConnectedSTAList = 19;
    TypeCmdSetMacAddress = 20;
    TypeRespSetMacAddress = 21;
    TypeCmdSetPowerSaveMode = 22;
    TypeRespSetPowerSaveMode = 23;
    TypeCmdGetPowerSaveMode = 24;
    TypeRespGetPowerSaveMode = 25;
}

message EspHostedConfigPayload {
    EspHostedConfigMsgType msg = 1;
    oneof payload {
        EspHostedCmdGetStatus cmd_get_mac_address = 10;
        EspHostedRespGetStatus resp_get_mac_address = 11;
        EspHostedCmdGetStatus cmd_get_wifi_mode = 12;
        EspHostedRespGetStatus resp_get_wifi_mode = 13;
        EspHostedCmdGetStatus cmd_set_wifi_mode = 14;
        EspHostedRespGetStatus resp_set_wifi_mode = 15;
        EspHostedCmdConfig cmd_get_ap_config = 16;
        EspHostedRespConfig resp_get_ap_config = 17;
        EspHostedCmdConfig cmd_set_ap_config = 18;
        EspHostedRespConfig resp_set_ap_config = 19;
        EspHostedCmdConfig cmd_get_softap_config = 20;
        EspHostedRespConfig resp_get_softap_config = 21;
        EspHostedCmdConfig cmd_set_softap_config = 22;
        EspHostedRespConfig resp_set_softap_config = 23;
        EspHostedCmdGetStatus cmd_disconnect_ap = 24;
        EspHostedRespGetStatus resp_disconnect_ap = 25;
        EspHostedCmdScanResult cmd_scan_ap_list = 26;
        EspHostedRespScanResult resp_scan_ap_list = 27;
        EspHostedCmdConnectedSTA cmd_connected_stas_list = 28;
        EspHostedRespConnectedSTA resp_connected_stas_list = 29;
        EspHostedCmdSetMacAddress cmd_set_mac_address = 30;
        EspHostedRespSetMacAddress resp_set_mac_address = 31;
        EspHostedCmdSetPowerSaveMode cmd_set_power_save_mode = 32;
        EspHostedRespSetPowerSaveMode resp_set_power_save_mode = 33;
        EspHostedCmdGetPowerSaveMode cmd_get_power_save_mode = 34;
        EspHostedRespGetPowerSaveMode resp_get_power_save_mode = 35;
    }
}
